### ISSUE TITLE
MIRI build in CI is failing

### ISSUE THREAD

**Issue by striezel** [mvdnes/spin-rs#134]
See e. g. https://github.com/mvdnes/spin-rs/actions/runs/3987644011/jobs/6837740845.

```
 stderr:
error: Undefined Behavior: not granting access to tag <3224> because that would remove [SharedReadOnly for <3170>] which is strongly protected because it is an argument of call 814
   --> /home/runner/work/spin-rs/spin-rs/src/rwlock.rs:536:32
    |
536 |                 data: unsafe { &mut *inner.data.get() },
    |                                ^^^^^^^^^^^^^^^^^^^^^^ not granting access to tag <3224> because that would remove [SharedReadOnly for <3170>] which is strongly protected because it is an argument of call 814
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <3224> was created by a SharedReadWrite retag at offsets [0x8..0xc]
   --> /home/runner/work/spin-rs/spin-rs/src/rwlock.rs:536:38
    |
536 |                 data: unsafe { &mut *inner.data.get() },
    |                                      ^^^^^^^^^^^^^^^^
help: <3170> is this argument
   --> /home/runner/work/spin-rs/spin-rs/src/rwlock.rs:516:29
    |
516 |     fn try_upgrade_internal(self, strong: bool) -> Result<RwLockWriteGuard<'rwlock, T, R>, Self> {
    |                             ^^^^
    = note: BACKTRACE (of the first span):
    = note: inside `spin::rwlock::RwLockUpgradableGuard::<'_, i32>::try_upgrade_internal` at /home/runner/work/spin-rs/spin-rs/src/rwlock.rs:536:32: 536:54
note: inside `spin::rwlock::RwLockUpgradableGuard::<'_, i32>::try_upgrade`
   --> /home/runner/work/spin-rs/spin-rs/src/rwlock.rs:556:9
    |
556 |         self.try_upgrade_internal(true)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main::_doctest_main_src_rwlock_rs_545_0`
   --> src/rwlock.rs:549:7
    |
7   | match upgradeable.try_upgrade() {
    |       ^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main`
   --> src/rwlock.rs:553:3
    |
11  | } _doctest_main_src_rwlock_rs_545_0() }
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
```
_(There are more errors, but that should be enough to get the basic idea.)_

The failure seems to be related to [stacked borrows](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md), but checking violations of stacked borrows with MIRI is still considered experimental at this moment.

One way to fix this would be to add the `-Zmiri-disable-stacked-borrows` flag for invokations of MIRI, but according to [its documentation](https://github.com/rust-lang/miri#miri--z-flags-and-environment-variables) that is unsound:

> * `-Zmiri-disable-stacked-borrows` disables checking the experimental
  [Stacked Borrows] aliasing rules.  This can make Miri run faster, but it also
  means no aliasing violations will be detected.  Using this flag is **unsound**
  (but the affected soundness rules are experimental).

**Comment by striezel** [comment]
> One way to fix this would be to add the `-Zmiri-disable-stacked-borrows` flag for invokations of MIRI, but according to [its documentation](https://github.com/rust-lang/miri#miri--z-flags-and-environment-variables) that is unsound:
> 
> > * `-Zmiri-disable-stacked-borrows` disables checking the experimental
> >   [Stacked Borrows] aliasing rules.  This can make Miri run faster, but it also
> >   means no aliasing violations will be detected.  Using this flag is **unsound**
> >   (but the affected soundness rules are experimental).

If that is an acceptable solution for you, then the pull request #135 will fix those errors.

**Comment by zesterer** [comment]
I think it would be better to fix this properly, but I'm likely not going to have the time to do this for a week or more unfortunately.

**Comment by striezel** [comment]
> I think it would be better to fix this properly

Indeed.

However, the stacked borrows rules that trigger the error and thus cause the build to fail are still experimental. Therefore, I am not sure whether this error really indicates an actual problem or whether it is just a false positive. :thinking:

**Comment by zesterer** [comment]
This has been resolved in #136. In the future I want to look into exactly how fast and loose we can go with the stacked borrow rules to maximise the compiler's ability to optimise (right now, for example, each use of `Deref`/`DerefMut` effectively causes the compiler to drop all assumptions about the state of the underlying data by dereferencing a raw pointer each time). At least for now, the crate appears to be compliant with stacked borrows (at least, that which tests cover - which should be almost everything that matters).
